### ISSUE TITLE
[Enhancement] Have a true sync API

### ISSUE THREAD

**Issue by eMerzh** [nextcloud/bookmarks#2]
Like delicious, Semantic Scuttle or anything else.. but let's  a way to sync bookmarks

**Comment by rakshazi** [comment]
Hello, guys!
Any progress on this?
I just want create Chrome extension with chrome.bookmarks functionality, this means that OwnCloud Bookmarks will be synchronized with native chrome bookmarks (two-way sync in plan), but I need simple API for this: get full list of bookmarks and add new bookmarks to the OwnCloud server, may be it's ready already? Can you explain me, how can I get full list of bookmarks and add new bookmarks via API?

**Comment by jancborchardt** [comment]
@blizzz is there an API already? If not, might be good to add this. :)

**Comment by blizzz** [comment]
There are a few routes that are used by the bookmarklet as well, but that's not enough for syncing, as they don't report modifications dates e.g.

**Comment by rakshazi** [comment]
@blizzz when real API will be available?
Bookmarks is very good app, but it costs nothing if there is no sync, I very want to use OwnCloud Bookmarks instead of Chrome sync bookmarks, but can't do this without API :(
Please, do it ASAP.

**Comment by blizzz** [comment]
@rakshazi  ASAP? There are a million things I want to do, but a day is short on this planet ;) Any chance you make it to the [conference](https://owncloud.org/conf/) and we gonna have a look at it?

**Comment by rakshazi** [comment]
@blizzz sorry, but my english is very bad and I have no chance to go to the conference. May be I can help and add API controllers via pull request? I don't know OwnCloud API and functions, but if you allow me to do it myself, I will try to do this.

**Comment by blizzz** [comment]
> but my english is very bad 

Can be only improved by practicing, i.e. an excellent opportunity :) Also many are non-native speakers either. Maybe next time?

> May be I can help and add API controllers via pull request? I don't know OwnCloud API and functions, but if you allow me to do it myself, I will try to do this.

Absolutely. 

The controllers are the interface to whatever you want to expose it to  (like API). Requests to them are directed by routes as in https://github.com/owncloud/bookmarks/blob/master/appinfo/routes.php

The business logic in the end is located at controller/lib/ (which is acutally not the right spot and we should relocate this at a later point). You can use and extend bookmarks.php unless you see need for a new class.

What I assume you mostly need is DB access. Study code and read https://doc.owncloud.org/server/8.1/developer_manual/app/database.html, this should give you the pointers. In general there is the app dev documentation here, if you come across hurdles: https://doc.owncloud.org/server/8.1/developer_manual/app/index.html

I hope that is material for a start. But first, make a plan for the API (what should it do/offer) and ideally implement an open protocol that is widely used for bookmarks, so it's easier to migrate from here to there and vice versa.

Also, I'll try to answer all your questions.

**Comment by rakshazi** [comment]
@blizzz Thank you!
pS: I'm trying to develop OwnCloud Pocket app (just use Pocket API, php wrapper for it is ready) to study  how can I code for OwnCloud, after that (with some exp) I'll start working on API for OwnCloud Bookmarks

**Comment by etiess** [comment]
@rakshazi Good idea! But why not working on wallabag (www.wallabag.org) ? The collaboration is discussed here: https://github.com/wallabag/wallabag/issues/408

**Comment by rakshazi** [comment]
@etiess it's just project for studying OwnCloud, not for production at all :)

**Comment by rakshazi** [comment]
@blizzz please, help me with some... bug(?). In public controller I wrote following code:
```php
<?php
public function update($user = null, $password = null, $id = null, $url, $title, $tags = array(), $description, $is_public = false)
        {
                if(!$this->isUser($user, $password))
                {
                        return $this->newJsonErrorMessage("User could not be identified");
                }

                Bookmarks::editBookmark($user,$this->db, $id, $url, $title, $tags, $description, $is_public);

                return new JSONResponse(array('status' => 'ok', 'message' => 'Bookmark has been updated. I hope.'));

        }
```

And added the following in routes.php:
```php
<?php
array('name' => 'public#update', 'url' => '/public/rest/v1/bookmarks/{id}', 'verb' => 'PUT')
```

But when I try to do request (all data correct), I got following message: 
```json
{
  "message": "Current user is not logged in"
}
```
and HTTP status code 401. Explain please, what I doing wrong?

Thank you!

pS: API will be ready after fix of this problem. For now, it works only with bookmarks. No tags available now.

**Comment by rakshazi** [comment]
@blizzz do you have any ideas about it?

**Comment by blizzz** [comment]
I will look into it as soon as i can (which will probably be sometime between 
today and thursday evening).

**Comment by rakshazi** [comment]
@blizzz sorry for disturb, but can you help me with requested problem? I just want to do API ASAP

**Comment by blizzz** [comment]
@rakshazi The user is not authenticated for some reason. What one hand is missing that you should not require the admin, but I think there's still something else not quite right. See https://doc.owncloud.org/server/8.1/developer_manual/app/controllers.html#authentication

The `@CORS` annotation is probably missing as well – might be the real cause?

>  The only pieces that need to be changed are the annotations which disable the CSRF check (not needed for a REST call usually) and add support for CORS so your API can be accessed from other webapps. [source](https://doc.owncloud.org/server/8.1/developer_manual/app/tutorial.html#adding-a-restful-api-optional)

**Comment by blizzz** [comment]
I am away for two weeks, maybe @MorrisJobke could help you with that if you are still stuck after my hints.

**Comment by rakshazi** [comment]
@blizzz thank you! I'll try this variant. Have a nice day!
@MorrisJobke hello! Please, see the commit above and give me your opinion about solving for this problem. Thank you!

**Comment by wtimme** [comment]
Hey guys, it's good to get something going here and I love all the participation. Could we, however, stick to the topic and discuss sync API proposals?

No offense, @rakshazi. It's just that the last few notifications that I got were a bit too much of an support-and-help-me kinda way. That's not a bad thing in general, of course. I just feel that we're getting a little off-topic.

Thank you, guys, and have a great weekend!

**Comment by rakshazi** [comment]
@wtimme hi!
Sorry for this, but I don't know how can I write only to needed developer, because GitHub have no private messages, that's why I ask question about API in issue about API. Do you know the better way?

Best regrads, NIkita Kushnir.

**Comment by MorrisJobke** [comment]
> Sorry for this, but I don't know how can I write only to needed developer, because GitHub have no private messages, that's why I ask question about API in issue about API. Do you know the better way?

Separate ticket for your sepecific issue (in your case a PR that has a problem would also help developers to check it out and reproduce your problem) or IRC (#owncloud-dev on freenode).

Can you do a PR with your changes that aren't working?

**Comment by rakshazi** [comment]
@MorrisJobke what is PR? Sorry, but I don't know what this means

**Comment by MorrisJobke** [comment]
> @MorrisJobke what is PR? Sorry, but I don't know what this means

Oh sorry: PR means Pull Request. So you do some changes in your fork of this app and then propose a PR that contains your changes https://help.github.com/articles/using-pull-requests/

**Comment by rakshazi** [comment]
Ok, thank you for explanation, I'll do this later today or tomorrow 

Morris Jobke писал 07.09.2015 12:25: 

>> @MorrisJobke [1] what is PR? Sorry, but I don't know what this means
> 
> Oh sorry: PR means Pull Request. So you do some changes in your fork of this app and then propose a PR that contains your changes https://help.github.com/articles/using-pull-requests/ [2] 
> 
> --
> Reply to this email directly or view it on GitHub [3].

-- 

Nikita Kushnir
Support Engineer at OpsWAY.com 

24x7 Kaizen Magento support.
We ❤️ a lot of small changes. Daily. 

+373 68 82 20 25
nikus@opsway.com
skype: nixischev
telegram: rakshazi 

Links:
------
[1] https://github.com/MorrisJobke
[2] https://help.github.com/articles/using-pull-requests/
[3]
https://github.com/owncloud/bookmarks/issues/2#issuecomment-138249396

**Comment by marcelklehr** [comment]
Good news! The new API will be probably be in the next release: #338 and I already have a webextension ready that will work with it to sync bookmarks from the browser with the app: https://github.com/marcelklehr/floccus ;)

**Comment by marcelklehr** [comment]
Fixed by #338
